### PR TITLE
Remove remaining support for denormalised tokens in the ring.

### PR DESCRIPTION
**Pull request by pstibrany**
This PR removes remaining support for denormalised tokens in the ring.

Targeted for post 0.6 release.

Reason for this change is to simplify the code. Nobody should be using denormalised tokens anymore, we have provided smooth upgrade path (post 0.4.0 Cortex only writes normalised tokens).

As we don't expect to run a mix of old (with old fields in ring protobuf) and new (with some new field with the same number) Cortex versions in practice, I've removed reserved fields from ring protobuf and also didn't reserve number for just-removed `Tokens` field.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ring/ring.proto [hunk 1]
```diff
@@ -9,7 +9,9 @@ option (gogoproto.unmarshaler_all) = true;
 
 message Desc {
 	map<string,IngesterDesc> ingesters = 1 [(gogoproto.nullable) = false];
-	repeated TokenDesc tokens = 2 [(gogoproto.nullable) = false];
+
+	// Denormalised tokens used field number 2. We don't reserve this field, since we don't expect to run a mix
```
Comment (jtlisi) [inline pkg/ring/ring.proto]: IMO, If we aren't going to reserve the field, there is no reason for the comment.

File: pkg/ring/model.go [hunk 2]
```diff
@@ -256,7 +218,6 @@ func (d *Desc) Merge(mergeable memberlist.Mergeable, localCAS bool) (memberlist.
 
 	// Keep ring normalized.
```
Comment (jtlisi) [inline pkg/ring/model.go]: You should be able to remove this comment

File: pkg/ring/ring.proto (hunk 1 above)
Comment (pstibrany) [reply pkg/ring/ring.proto]: Right. I wanted to add that comment somewhere, but wasn't sure where. I've now also removed previously reserved fields, for the same reason.

File: pkg/ring/ring.proto [hunk 3]
```diff
@@ -9,23 +9,15 @@ option (gogoproto.unmarshaler_all) = true;
 
 message Desc {
 	map<string,IngesterDesc> ingesters = 1 [(gogoproto.nullable) = false];
-	repeated TokenDesc tokens = 2 [(gogoproto.nullable) = false];
 }
 
 message IngesterDesc {
-	reserved 4, 5; // old, deprecated fields
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/ring.proto]: I've added this comment to the PR:

> As we don't expect to run a mix of old (with old fields in ring protobuf) and new (with some new field with the same number) Cortex versions in practice, I've removed reserved fields from ring protobuf and also didn't reserve number for just-removed Tokens field.

What do you think? It is little safer to keep, but the risk is negligible.

[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/ring/ring.proto]: Yeah, I've read the PR description. Given doesn't cause any harm keeping it, and it's just 1 line, and it's slightly safer... I would stay on the "better safe than sorry" side, so I would keep it. But I don't have strong feelings on it.

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/ring.proto]: OK, I'll put it back and will not try to be too smart about it. Better safe than sorry.

File: pkg/ring/model.go [hunk 4]
```diff
@@ -426,3 +373,25 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
...
 	}
 }
+
+type TokenDesc struct {
+	Token    uint32
+	Ingester string
+}
+
+// Returns sorted list of tokens with ingester names.
+func (d *Desc) migrateRing() []TokenDesc {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/model.go]: I know that #1764 calls it `GetNavigator`, but it also introduces entire new `TokenNavigator` type for it, with many more methods required for advanced tokens searches. For now, `getTokens` sounds good.

**Comment by pstibrany** [comment]
Thanks guys both for your reviews. I've addressed (hopefully all) your feedback.

> The cleanup LGTM. I've also manually tested the ingester transfer.

Btw, we have integration test doing this already (it found some bugs in my code in the past).

File: pkg/ring/ring.proto (hunk 1 above)
Comment (jtlisi) [reply pkg/ring/ring.proto]: I think the best place to note it would be in the changelog

**Pull request merged**